tools/osx-notarize.sh

```diff
@@ -1,34 +1,83 @@
 #!/bin/sh
 
-# Uses gon, from https://github.com/mitchellh/gon, to notarize a generated node-<version>.pkg file
-# with Apple for installation on macOS Catalina and later as validated by Gatekeeper.
+# Notarize a generated node-<version>.pkg file as an Apple requirement for installation on macOS Catalina and later, as validated by Gatekeeper.
+# Uses gon (Xcode version < 13.0) or notarytool (Xcode >= 13.0).
 
-set -e
-
-gon_version="0.2.2"
-gon_exe="${HOME}/.gon/gon_${gon_version}"
+version() {
+  echo "$@" | awk -F. '{ printf("%d%03d%03d%03d\n", $1,$2,$3,$4); }' || echo "0"
+}
 
+xcode_version=$(xcodebuild -version | awk '/Xcode/ {print $2}')
+xcode_version_result=$(version "$xcode_version")
+xcode_version_threshold=$(version "13.0")
 pkgid="$1"
 
-[ -z "$pkgid" ] && \
-  echo "Usage: $0 <pkgid>" \
+if [ -z "$pkgid" ]; then
+  echo "Usage: $0 <pkgid>"
   exit 1
+fi
 
 # shellcheck disable=SC2154
-[ -z "$NOTARIZATION_ID" ] && \
-  echo "No NOTARIZATION_ID environment var. Skipping notarization." \
+if [ -z "$NOTARIZATION_ID" ]; then
+  echo "No NOTARIZATION_ID environment variable. Skipping notarization."
   exit 0
+fi
 
-set -x
-
-mkdir -p "${HOME}/.gon/"
+if [ -z "$NOTARIZATION_PASSWORD" ]; then
+  echo "No NOTARIZATION_PASSWORD environment variable. Skipping notarization."
+  exit 0
+fi
 
-if [ ! -f "${gon_exe}" ]; then
-  curl -sL "https://github.com/mitchellh/gon/releases/download/v${gon_version}/gon_${gon_version}_macos.zip" -o "${gon_exe}.zip"
-  (cd "${HOME}/.gon/" && rm -f gon && unzip "${gon_exe}.zip" && mv gon "${gon_exe}")
+if [ -z "$NOTARIZATION_TEAM_ID" ]; then
+  echo "No NOTARIZATION_TEAM_ID environment variable. Skipping notarization."
+  exit 0
 fi
 
-sed -e "s/{{appleid}}/${NOTARIZATION_ID}/" -e "s/{{pkgid}}/${pkgid}/" tools/osx-gon-config.json.tmpl \
-  > gon-config.json
+if [ "$xcode_version_result" -lt "$xcode_version_threshold" ]; then
+  echo "Notarization process is done with gon."
+  set -x
+
+  gon_version="0.2.2"
+  gon_exe="${HOME}/.gon/gon_${gon_version}"
+
+  mkdir -p "${HOME}/.gon/"
+
+  if [ ! -f "${gon_exe}" ]; then
+    curl -sL "https://github.com/mitchellh/gon/releases/download/v${gon_version}/gon_${gon_version}_macos.zip" -o "${gon_exe}.zip"
+    (cd "${HOME}/.gon/" && rm -f gon && unzip "${gon_exe}.zip" && mv gon "${gon_exe}")
+  fi
+
+  sed -e "s/{{appleid}}/${NOTARIZATION_ID}/" -e "s/{{pkgid}}/${pkgid}/" tools/osx-gon-config.json.tmpl \
+    > gon-config.json
+
+  "${gon_exe}" -log-level=info gon-config.json
+
+else
+  echo "Notarization process is done with Notarytool."
+
+  if ! command -v xcrun notarytool > /dev/null
+  then
+      echo "Notarytool is not present in the system. Notarization has failed."
+      exit 1
+  fi
+
+  # Submit the package for notarization
+  notarization_output=$(
+    xcrun notarytool submit \
+      --apple-id "$NOTARIZATION_ID" \
+      --password "$NOTARIZATION_PASSWORD" \
+      --team-id "$NOTARIZATION_TEAM_ID" \
```

Comment:
Could we add a TODO here to use `--keychain-profile` option?

Reply:
@tony-go for my benefit what is --keychain-profile to help me understand why we'd want to use it.

Reply:
> Could we add a TODO here to use --keychain-profile option?

Great point!, I will include it. I can add also another TODO for removing `gon` once macos10.15 is deprecated from the release infra.

Reply:
@mhdawson, this link (https://github.com/nodejs/node/pull/48701#discussion_r1257298703) can provide you with more context. I think it is safer because we are not injecting the secrets into the environment. We can generate a profile with the secrets included, so that the profile is available when running this script. However, in order to do that, we need to make changes to the Ansible scripts.

Reply:
@UlisesGascon great explanation :)

Reply:
@UlisesGascon thanks for the explanation. Once you add the TODO let me know so I can re-run the CI

Reply:
@mhdawson changes added in [076dbcb](https://github.com/nodejs/node/pull/48701/commits/076dbcb91647a51df3a8848ce031fbb4fde64f59)

Reply:
I added `NOTARIZATION_TEAM_ID` environmental secret to the [iojs+release config](https://ci-release.nodejs.org/job/iojs+release/configure), so it once this is merged we won't expect any environmental issues.